### ISSUE TITLE
contrib/contrib/go-chi: set the resource the actual path instead of "unknown" at 404 HTTP code

### ISSUE THREAD

**Issue by bitvirago** [DataDog/dd-trace-go#1198]
I started to use the chi trace middleware and I realized that when the chi gets an unknown path in the HTTP request the middleware set "unknown" in the Datadog resource because in this case the `RoutePattern()` gives back an empty string. My proposal is in this case the resource should be the actual path. 

the actual code snippet:
```
	// set the resource name as we get it only once the handler is executed
	resourceName := chi.RouteContext(r.Context()).RoutePattern()
	if resourceName == "" {
		resourceName = "unknown"
	}
```

the changes would be:
```
	// set the resource name as we get it only once the handler is executed
	resourceName := chi.RouteContext(r.Context()).RoutePattern()
	if resourceName == "" {
		resourceName = r.URL.Path
		if resourceName == "" {
			resourceName = "unknown"
		}
	}
```

Any idea or suggestion?

**Comment by taowata** [comment]
Hi,

I've encountered the same issue and I'm interested in working on it.

![スクリーンショット 2023-05-02 1 31 08](https://user-images.githubusercontent.com/57245344/235488764-56b6165e-3c3d-40c7-af3e-f61afe774eec.png)

Is anyone already working on this or has there been any progress? 
I think the solution proposed by @bitvirago could resolve this problem.
If it's okay, I'd like to create a PR to resolve this issue. Please let me know if I can help.

Thanks,

**Comment by taowata** [comment]
In my case, the issue occurred when a middleware performing request validation encountered an error.
Here's an example of the test code to reproduce the issue:
```go
import (
	"net/http"
	"net/http/httptest"
	"testing"

	chitrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi.v5"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/render"
	"github.com/stretchr/testify/assert"
)

func someValidationMiddleware() func(http.Handler) http.Handler {
	return func(h http.Handler) http.Handler {
		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
			if somethingWrong := true; somethingWrong {
				render.Status(r, http.StatusBadRequest)
				render.Respond(rw, r, render.M{"status": "error"})
				return
			}
			h.ServeHTTP(rw, r)
		})
	}
}

func TestUnknown(t *testing.T) {
	assert := assert.New(t)
	mt := mocktracer.Start()
	defer mt.Stop()

	router := chi.NewRouter()
	router.Use(chitrace.Middleware(), someValidationMiddleware())
	router.Post("/user/123", func(w http.ResponseWriter, r *http.Request) {})

	r := httptest.NewRequest(http.MethodPost, "/user/123", nil)
	w := httptest.NewRecorder()
	router.ServeHTTP(w, r)
	response := w.Result()
	defer response.Body.Close()
	spans := mt.FinishedSpans()

	assert.Equal(400, response.StatusCode)
	assert.Equal("POST /user/123", spans[0].Tag("resource.name")) // not POST /user/123 but POST unknown
}
```

**Comment by taowata** [comment]
In my case, the issue occurred when a middleware performing request validation encountered an error.
Here's an example of the test code to reproduce the issue:
```go
import (
	"net/http"
	"net/http/httptest"
	"testing"

	chitrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi.v5"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/render"
	"github.com/stretchr/testify/assert"
)

func someValidationMiddleware() func(http.Handler) http.Handler {
	return func(h http.Handler) http.Handler {
		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
			if somethingWrong := true; somethingWrong {
				render.Status(r, http.StatusBadRequest)
				render.Respond(rw, r, render.M{"status": "error"})
				return
			}
			h.ServeHTTP(rw, r)
		})
	}
}

func TestUnknown(t *testing.T) {
	assert := assert.New(t)
	mt := mocktracer.Start()
	defer mt.Stop()

	router := chi.NewRouter()
	router.Use(chitrace.Middleware(), someValidationMiddleware())
	router.Post("/user/123", func(w http.ResponseWriter, r *http.Request) {})

	r := httptest.NewRequest(http.MethodPost, "/user/123", nil)
	w := httptest.NewRecorder()
	router.ServeHTTP(w, r)
	response := w.Result()
	defer response.Body.Close()
	spans := mt.FinishedSpans()

	assert.Equal(400, response.StatusCode)
	assert.Equal("POST /user/123", spans[0].Tag("resource.name")) // not POST /user/123 but POST unknown
}
```

**Comment by taowata** [comment]
This issue could also be solved in a more flexible way by allowing customization of the resouceName, as suggested in #1950, leaving it up to the user.

**Comment by zarirhamza** [comment]
We prefer not to expose the path by default due to policy. I'm keeping the PR open if you want to discuss other options for default behavior for "unknown" but we are merging #1950 soon which as you mentioned provides a more generic solution for this.

**Comment by taowata** [comment]
@zarirhamza 
Thank you for your feedback. I understand the concern about exposing the path by default due to policy and closed the PR.

In the meantime, I have created a new PR #2024 to apply the solution from #1950 to chi.v5, which I believe aligns more closely with your direction. I'd greatly appreciate if you could take some time to review this new PR.

**Comment by zarirhamza** [comment]
Hello! Thank you both for your help and we know support `WithResourceNamer` for both go-chi and go-chi.v5. Closing this issue now as everything is solved, but feel free to open another issue in the future with any other concerns.
